File: app/src/main/java/com/qingshuige/tangyuan/SearchActivity.java
package com.qingshuige.tangyuan;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.qingshuige.tangyuan.network.ApiHelper;
import com.qingshuige.tangyuan.network.Comment;
import com.qingshuige.tangyuan.network.PostMetadata;
import com.qingshuige.tangyuan.network.User;
import com.qingshuige.tangyuan.viewmodels.CommentCardAdapter;
import com.qingshuige.tangyuan.viewmodels.CommentInfo;
import com.qingshuige.tangyuan.viewmodels.PostCardAdapter;
import com.qingshuige.tangyuan.viewmodels.PostInfo;
import com.qingshuige.tangyuan.viewmodels.UserCardAdapter;
import com.qingshuige.tangyuan.viewmodels.UserInfo;

import java.util.List;
import java.util.concurrent.CountDownLatch;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchActivity extends AppCompatActivity {

    private String keyword;

    private Toolbar toolbar;
    private ProgressBar pgBar;

    private RecyclerView rcvPost;
    private TextView textPostSearchTitle;
    private PostCardAdapter postAdapter;

    private RecyclerView rcvUser;
    private TextView textUserSearchTitle;
    private UserCardAdapter userAdapter;

    private RecyclerView rcvComment;
    private TextView textCommentSearchTitle;
    private CommentCardAdapter commentAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_search);
//        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
//            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
//            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
//            return insets;
//        });

        setSupportActionBar(findViewById(R.id.toolbar));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);

        keyword = getIntent().getStringExtra("keyword");

        toolbar = findViewById(R.id.toolbar);
        rcvPost = findViewById(R.id.rcvPost);
        pgBar = findViewById(R.id.pgBar);
        textPostSearchTitle = findViewById(R.id.textPostSearchTitle);
        rcvUser = findViewById(R.id.rcvUser);
        textUserSearchTitle = findViewById(R.id.textUserSearchTitle);
        rcvComment = findViewById(R.id.rcvComment);
        textCommentSearchTitle = findViewById(R.id.textCommentSearchTitle);

        toolbar.setTitle(getString(R.string.search) + ": " + keyword);

        DividerItemDecoration div = new DividerItemDecoration(this, DividerItemDecoration.VERTICAL);

        //Post
        postAdapter = new PostCardAdapter();
        postAdapter.setOnItemClickListener(this::startPostActivity);
        rcvPost.setAdapter(postAdapter);
        rcvPost.addItemDecoration(div);
        rcvPost.setLayoutManager(new LinearLayoutManager(this));

        //User
        userAdapter = new UserCardAdapter(this);
        rcvUser.setAdapter(userAdapter);
        rcvUser.setLayoutManager(new LinearLayoutManager(this));
        rcvUser.addItemDecoration(div);

        //Comment
        commentAdapter = new CommentCardAdapter();
        rcvComment.setAdapter(commentAdapter);
        rcvComment.setLayoutManager(new LinearLayoutManager(this));
        rcvComment.addItemDecoration(div);

        initializeUI();

    }

    private void initializeUI() {
        new Thread(() -> {
            CountDownLatch latch = new CountDownLatch(3);

            //帖子搜索结果
            TangyuanApplication.getApi().searchPostByKeyword(keyword).enqueue(new Callback<List<PostMetadata>>() {
                @Override
                public void onResponse(Call<List<PostMetadata>> call, Response<List<PostMetadata>> response) {
                    if (response.code() == 200 && response.body() != null) {
                        runOnUiThread(() -> textPostSearchTitle.setText(textPostSearchTitle.getText() + ": " + response.body().size()));
                        ApiHelper.getPostInfoByMetadataFastAsync(response.body(), result -> {
                            if (result != null) {
                                runOnUiThread(() -> {
                                    postAdapter.replaceDataSet(result);
                                });
                            }
                        });
                    } else {
                        runOnUiThread(() -> {
                            textPostSearchTitle.setText(R.string.no_post_result);
                        });
                    }
                    latch.countDown();
                }

                @Override
                public void onFailure(Call<List<PostMetadata>> call, Throwable throwable) {
                    runOnUiThread(() -> {
                        textPostSearchTitle.setText(textPostSearchTitle.getText() + ": " + getString(R.string.network_error));
                    });
                }
            });

            //用户搜索结果
            TangyuanApplication.getApi().searchUserByKeyword(keyword).enqueue(new Callback<List<User>>() {
                @Override
                public void onResponse(Call<List<User>> call, Response<List<User>> response) {
                    if (response.code() == 200 && response.body() != null) {
                        List<User> users = response.body();
                        runOnUiThread(() -> textUserSearchTitle.setText(textUserSearchTitle.getText() + ": " + response.body().size()));
                        ApiHelper.getInfoFastAsync(users, new ApiHelper.UserInfoConstructor(), result -> {
                            if (result != null) {
                                runOnUiThread(() -> {
                                    userAdapter.replaceDataset(result);
                                });
                            }

                        });
                    } else {
                        runOnUiThread(() -> {
                            textUserSearchTitle.setText(R.string.no_user_result);
                        });
                    }
                    latch.countDown();
                }

                @Override
                public void onFailure(Call<List<User>> call, Throwable throwable) {

                }
            });

            //评论搜索结果
            TangyuanApplication.getApi().searchCommentByKeyword(keyword).enqueue(new Callback<List<Comment>>() {
                @Override
                public void onResponse(Call<List<Comment>> call, Response<List<Comment>> response) {
                    if (response.code() == 200 && response.body() != null) {
                        List<Comment> comments = response.body();
                        runOnUiThread(() -> textCommentSearchTitle.setText(textCommentSearchTitle.getText() + ": " + response.body().size()));
                        ApiHelper.getInfoFastAsync(comments, new ApiHelper.CommentInfoConstructor(), result -> {
                            runOnUiThread(() -> commentAdapter.replaceDataset(result));
                            latch.countDown();
                        });
                    } else {
                        runOnUiThread(() -> {
                            textCommentSearchTitle.setText(R.string.no_comment_result);
                        });
                    }
                }

                @Override
                public void onFailure(Call<List<Comment>> call, Throwable throwable) {

                }
            });

            try {
                latch.await();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }

            runOnUiThread(() -> pgBar.setVisibility(View.GONE));

        }).start();
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }

    private void startPostActivity(int postId) {
        Intent intent = new Intent(this, PostActivity.class);
        intent.putExtra("postId", postId);
        startActivity(intent);
    }
}

File: app/src/main/java/com/qingshuige/tangyuan/SearchActivity.kt
package com.qingshuige.tangyuan

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
//import androidx.activity.EdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.qingshuige.tangyuan.network.ApiHelper
import com.qingshuige.tangyuan.network.Comment
import com.qingshuige.tangyuan.network.PostMetadata
import com.qingshuige.tangyuan.network.User
import com.qingshuige.tangyuan.viewmodels.CommentCardAdapter
import com.qingshuige.tangyuan.viewmodels.PostCardAdapter
import com.qingshuige.tangyuan.viewmodels.UserCardAdapter
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.concurrent.CountDownLatch
import kotlin.concurrent.thread

class SearchActivity : AppCompatActivity() {

    private lateinit var keyword: String
    private lateinit var toolbar: Toolbar
    private lateinit var pgBar: ProgressBar
    private lateinit var rcvPost: RecyclerView
    private lateinit var textPostSearchTitle: TextView
    private lateinit var postAdapter: PostCardAdapter
    private lateinit var rcvUser: RecyclerView
    private lateinit var textUserSearchTitle: TextView
    private lateinit var userAdapter: UserCardAdapter
    private lateinit var rcvComment: RecyclerView
    private lateinit var textCommentSearchTitle: TextView
    private lateinit var commentAdapter: CommentCardAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        EdgeToEdge.enable(this)
        setContentView(R.layout.activity_search)
//        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
//            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
//            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
//            return insets;
//        });

        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }

        keyword = intent.getStringExtra("keyword") ?: ""

        toolbar = findViewById(R.id.toolbar)
        rcvPost = findViewById(R.id.rcvPost)
        pgBar = findViewById(R.id.pgBar)
        textPostSearchTitle = findViewById(R.id.textPostSearchTitle)
        rcvUser = findViewById(R.id.rcvUser)
        textUserSearchTitle = findViewById(R.id.textUserSearchTitle)
        rcvComment = findViewById(R.id.rcvComment)
        textCommentSearchTitle = findViewById(R.id.textCommentSearchTitle)

        toolbar.title = "${getString(R.string.search)}: $keyword"

        val div = DividerItemDecoration(this, DividerItemDecoration.VERTICAL)

        //Post
        postAdapter = PostCardAdapter()
        postAdapter.setOnItemClickListener { postId -> startPostActivity(postId, 0) }
        rcvPost.adapter = postAdapter
        rcvPost.addItemDecoration(div)
        rcvPost.layoutManager = LinearLayoutManager(this)

        //User
        userAdapter = UserCardAdapter(this)
        userAdapter.setOnUserClickListener { info -> startUserActivity(info.user.userId) }
        rcvUser.adapter = userAdapter
        rcvUser.layoutManager = LinearLayoutManager(this)
        rcvUser.addItemDecoration(div)

        //Comment
        commentAdapter = CommentCardAdapter()
        commentAdapter.setReplyButtonVisible(false)
        commentAdapter.setOnAvatarClickListener { info -> startUserActivity(info.userId) }
        commentAdapter.setOnTextClickListener { info -> startPostActivity(info.comment?.postId!!, info.commentId) }
        rcvComment.adapter = commentAdapter
        rcvComment.layoutManager = LinearLayoutManager(this)
        rcvComment.addItemDecoration(div)

        initializeUI()
    }

    private fun startUserActivity(userId: Int) {
        val intent = Intent(this, UserActivity::class.java)
        intent.putExtra("userId", userId)
        startActivity(intent)
    }

    private fun initializeUI() {
        thread {
            val latch = CountDownLatch(3)

            //帖子搜索结果
            TangyuanApplication.getApi().searchPostByKeyword(keyword).enqueue(object : Callback<List<PostMetadata>> {
                override fun onResponse(call: Call<List<PostMetadata>>, response: Response<List<PostMetadata>>) {
                    if (response.code() == 200 && response.body() != null) {
                        runOnUiThread { 
                            textPostSearchTitle.text = "${textPostSearchTitle.text}: ${response.body()!!.size}" 
                        }
                        ApiHelper.getInfoFastAsync(response.body()!!, ApiHelper.PostInfoConstructor()) { result ->
                            result?.let {
                                runOnUiThread { postAdapter.replaceDataSet(it) }
                            }
                        }
                    } else {
                        runOnUiThread {
                            textPostSearchTitle.setText(R.string.no_post_result)
                        }
                    }
                    latch.countDown()
                }

                override fun onFailure(call: Call<List<PostMetadata>>, throwable: Throwable) {
                    runOnUiThread {
                        textPostSearchTitle.text = "${textPostSearchTitle.text}: ${getString(R.string.network_error)}"
                    }
                }
            })

            //用户搜索结果
            TangyuanApplication.getApi().searchUserByKeyword(keyword).enqueue(object : Callback<List<User>> {
                override fun onResponse(call: Call<List<User>>, response: Response<List<User>>) {
                    if (response.code() == 200 && response.body() != null) {
                        val users = response.body()!!
                        runOnUiThread { 
                            textUserSearchTitle.text = "${textUserSearchTitle.text}: ${users.size}" 
                        }
                        ApiHelper.getInfoFastAsync(users, ApiHelper.UserInfoConstructor()) { result ->
                            result?.let {
                                runOnUiThread {
                                    userAdapter.replaceDataset(it)
                                }
                            }
                        }
                    } else {
                        runOnUiThread {
                            textUserSearchTitle.setText(R.string.no_user_result)
                        }
                    }
                    latch.countDown()
                }

                override fun onFailure(call: Call<List<User>>, throwable: Throwable) {}
            })

            //评论搜索结果
            TangyuanApplication.getApi().searchCommentByKeyword(keyword).enqueue(object : Callback<List<Comment>> {
                override fun onResponse(call: Call<List<Comment>>, response: Response<List<Comment>>) {
                    if (response.code() == 200 && response.body() != null) {
                        val comments = response.body()!!
                        runOnUiThread { 
                            textCommentSearchTitle.text = "${textCommentSearchTitle.text}: ${comments.size}" 
                        }
                        ApiHelper.getInfoFastAsync(comments, ApiHelper.CommentInfoConstructor()) { result ->
                            runOnUiThread { commentAdapter.replaceDataset(result!!) }
                        }
                    } else {
                        runOnUiThread {
                            textCommentSearchTitle.setText(R.string.no_comment_result)
                        }
                    }
                    latch.countDown()
                }

                override fun onFailure(call: Call<List<Comment>>, throwable: Throwable) {}
            })

            try {
                latch.await()
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }

            runOnUiThread { pgBar.visibility = View.GONE }
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    private fun startPostActivity(postId: Int, commentId: Int) {
        val intent = Intent(this, PostActivity::class.java)
        intent.putExtra("postId", postId)
        intent.putExtra("commentId", commentId)
        startActivity(intent)
    }
}